Update Contributing file (#407)
PlaceholderAPI/PlaceholderAPI: .github/CONTRIBUTING.md

[issue]: https://github.com/PlaceholderAPI/PlaceholderAPI/issues/new
[discord]: https://helpch.at/discord

> [Code of Conduct]: https://github.com/PlaceholderAPI/PlaceholderAPI/blob/master/CODE_OF_CONDUCT.md

Author: Even tho it's case insensitive do I always try to match the text that this is used in... 🤷 
Just a habit of mine.

[wiki]: https://github.com/PlaceholderAPI/PlaceholderAPI/blob/master/wiki

# Contributing Guidelines
We welcome everyone to contribute towards the PlaceholderAPI Project, but doing so will require you to follow specific rules to keep a consistent and welcoming way of contributing.

If you have any questions about this contributing file, feel free to contact us on our [Discord Server][discord].

## Issues
Like any other Project can you encounter bugs or a feature is missing for you in the plugin.  

> For that can you open an [issue] to report a bug or suggest a new feature to be added.

Author: I don't see where this could be seen as a question... Like it tells you that you can open an issue for reporting bugs or suggesting features.
It's not like "Can you please open an issue to do this?" but rather "Please open an issue for this" in my eyes... But you know... Everyone reads different.

> When doing so, make sure you follow the below rules.

Author: This honestly sounds odd when you read it out loud.

> When doing so, make sure you follow the below rules.

Reviewer: I agree with the suggested changes

### Follow the Template
We have Issue templates to help us get the required information more easly. Please follow the provided template when either filing a bug report or feature request.  

> We reserve the right to close your issue without warning for not following the templates.

Reviewer 2: Not a fan of the word "right" in this context. Perhaps consider the following:
```suggestion
Your issue may be closed without warning for not following the template.
```

Although, I'd actually say that issues may be closed for any reason, not necessarily related to the template.

> ### Using Latest Version

Reviewer 2: ```suggestion
### Use the Latest Version
```

> When it comes to bug reports should you always check first, if you're using the latest release of PlaceholderAPI.  

Reviewer 2: ```suggestion
When it comes to bug reports should you always check first, that you're using the latest release of PlaceholderAPI.  
```

> Often is a bug you encounter fixed in a newer release.

Reviewer 2: ```suggestion
Often the bug you've encountered, is fixed in a newer version.
```

Might not be a release, but rather a dev build.

> You can optionally also try out development builds to see, if your bug is fixed in those.

Reviewer 2: See above. This line can be merged into one

> The same rules apply for when making a feature request.

Reviewer 2: ```suggestion
The same rules apply when making a feature request.
```
Concise language

> ### No duplicate issue

Reviewer 2: Why do ### titles suddenly stop being capitalised here? Not that it's bad, but it should be consistent. Decide on a single style. Personally, I prefer this form, with only the first word capitalised.

Author: Most likely due to my german grammar knowledge where words prefixed with "the" (Nouns I suppose) are always capitalized.

> Make sure that there isn't any existing issue, which is still open, or is closed with a solution/explanation.  

Reviewer 2: ```suggestion
Make sure that there aren't any existing issues relating to the problem, which are still open, or are closed with a solution/explanation.  
```
Fix tenses

Opening a separate issue for a bug report or feature request, that already exists on the issue tracker only slows down the process of fixing the bug or implementing the feature.

If an issue with the bug or feature you want to report/suggest exists, comment on it with your info (bug reports) or give it a :thumbsup: (Feature Request) to show that this is important for you.

### Issue isn't caused by external source

> PlaceholderAPI provides a feature to have extensions (separate jar files) for placeholders. This gives it a possability that an issue you encounter is caused by said extensions or a separate plugin that uses those extensions.  

Reviewer 2: Yeah, they're extensions, but be consistent with your naming. Call them expansions, not extensions.

> In those cases should you report the issue to the issue tracker of the extension or plugin.

Reviewer 2: See above, regarding to the naming

## Pull requests

> As an open source project are we welcoming all contributions to improve PlaceholderAPI, being it changes to its code like bug fixes or new features, or contributions to its documentation such as the Wiki or the Javadoc.

Reviewer 2: Might be worth linking wiki & javadoc here, and making them non-capitalised.

Author: Where are the Javadocs tho? There is to my knowledge still not a online version of it viewable.

### Code contributions

> Any contributions to PlaceholderAPI's code should be done towards the `develop` branch. Targeting the `master` branch in your Pull request may get it closed without warning.  

Reviewer 2: I believe this isn't actually the case. Consult with lemmo

Author: See #418 which suggest the branches to be setup.
Right now do only the `master` and `docs/wiki` (Will try to rename it to `docs` later on) as actual branches exist, so develop has to be created first.
But Lemmo suggested that master and develop should be protected branches and that master should only accept PRs from develop as it would be reserved for official releases of PlaceholderAPI.

> Additionally should you follow the `Google Codestyle` when changing the code and Javadoc.

Reviewer 2: No, this isn't the case. PAPI uses it's own codestyle, which is in the config folder iirc.

Author: It uses a codestyle that is adabted from the Google Codestyle, which @extendedclip once suggested to be added (And caused this huge debate with Sxtanna)

Some noteworthy points about Javadoc:

- `@param` tags are in a single line and any additional param tag after one don't have an empty line in between.  
Example:  
```java
/**
 * @param param1 description
 * @param param2 description
 */
```
- Tags follow the order `@param`, `@return`, `@since` and `@deprecated`

> - The `@deprecated` tag should mention the version of when the annotated object will be removed. This usually would be the next minor version.

Reviewer 2: I believe all of this is handled by the codestyle file.

### Wiki/Javadoc contributions

> If your Pull request only targets the [wiki] or only changes the Javadoc of PlaceholderAPI (And not its actual code) should Pull requests target the `docs` branch.  

Reviewer 2: ```suggestion
If your pull request only targets the [wiki] or only changes the javadocs of PlaceholderAPI (and not its actual code), it should be to the `docs` branch.
```

How do you change the javadocs without editing the actual source files though? Wouldn't that be in a non-docs branch? Or are the source files mirrored to the docs branch, exclusively for this purpose?

Author: If the changes are affecting the actual code would it be on the develop branch.
When you, however, only edit the docs to f.e. fix typos or invalid tags, should it target the docs branch which at a later point could be merged into develop.
The point of this would be to prevent an extreme amount of "fix typo" PRs that could trigger new builds with no changes other than Javadoc corrections.

Author: Also... "it should be to the [...]"? That sounds horrible. I think "it should be towards the [...]" would be a better solution here.

> The develop branch should only be targeted when you also change some of PlaceholderAPI's code.

Reviewer 2: See a few comments above, regarding consulting lemmo.

Author: See my response about this...

## Code of Conduct
We have a [Code of Conduct] to maintain a welcoming atmosphere in this project.  

> If your contributions go against the above linked Code of Conduct will we reserve the right t deny or revert your contributions and to permanently remove your access from this project.

Reviewer 2: ```suggestion
If your contributions go against the code of conduct, linked above, we reserve the right to deny or revert your contributions.
```

Author: I think `Code of Conduct` is fine in that capitalisation as I see no real reason why it should suddenly be all lowercase.
